packages/flutter_blue/tizen/src/bluetooth_device_controller.cc

```diff
@@ -0,0 +1,255 @@
+    return (is_connecting_ ? State::kConnecting : State::kDisconnecting);
+  } else {
+    bool is_connected = false;
+    int res = bt_device_is_profile_connected(address_.c_str(), BT_PROFILE_GATT,
+                                             &is_connected);
+    LOG_ERROR("bt_device_is_profile_connected", get_error_message(res));
+    return (is_connected ? State::kConnected : State::kDisconnected);
+  }
+}
+
+std::vector<proto::gen::BluetoothDevice>&
+BluetoothDeviceController::ProtoBluetoothDevices() noexcept {
+  return proto_bluetooth_devices_;
+}
+
+const std::vector<proto::gen::BluetoothDevice>&
+BluetoothDeviceController::cProtoBluetoothDevices() const noexcept {
+  return proto_bluetooth_devices_;
+}
+
+void BluetoothDeviceController::Connect(bool auto_connect) {
+  std::unique_lock lock(operation_mutex_);
+  auto_connect = false;  // TODO - fix. API fails when autoconnect==true
+  if (State() == State::kDisconnected) {
+    is_connecting_ = true;
+    int res = bt_gatt_connect(address_.c_str(), auto_connect);
+    LOG_ERROR("bt_gatt_connect", get_error_message(res));
+  } else {
+    std::string message =
+        "trying to connect to a device with State!=DISCONNECTED " +
+        std::to_string((int)State());
+    throw BtException(message);
+  }
+}
+
+void BluetoothDeviceController::Disconnect() {
+  std::scoped_lock lock(operation_mutex_);
+  auto st = State();
+  if (st == State::kConnected) {
+    services_.clear();
+    is_disconnecting_ = true;
+    int res = bt_gatt_disconnect(address_.c_str());
+    LOG_ERROR("bt_gatt_disconnect", get_error_message(res));
+  }
+}
+
+bt_gatt_client_h BluetoothDeviceController::GetGattClient(
+    const std::string& address) {
+  std::scoped_lock lock(gatt_clients_.mutex_);
+
+  auto it = gatt_clients_.var_.find(address);
+  bt_gatt_client_h client = nullptr;
+
+  if (it == gatt_clients_.var_.end()) {
+    int res = bt_gatt_client_create(address.c_str(), &client);
+    LOG_ERROR("bt_gatt_client_create", get_error_message(res));
+
+    if ((res == BT_ERROR_NONE || res == BT_ERROR_ALREADY_DONE) && client) {
+      gatt_clients_.var_.emplace(address, client);
+    } else
+      throw BtException(res, "bt_gatt_client_create");
+  } else {
+    client = it->second;
+  }
+  return client;
+}
+
+void BluetoothDeviceController::DestroyGattClientIfExists(
+    const std::string& address) noexcept {
+  std::scoped_lock lock(gatt_clients_.mutex_);
+  auto it = gatt_clients_.var_.find(address);
+  if (it != gatt_clients_.var_.end()) {
+    auto res = bt_gatt_client_destroy(it->second);
+    if (!res) gatt_clients_.var_.erase(address);
+    LOG_ERROR("bt_gatt_client_destroy", get_error_message(res));
+  }
+}
+
+void BluetoothDeviceController::DiscoverServices() {
+  std::scoped_lock lock(operation_mutex_);
+
+  struct Scope {
+    BluetoothDeviceController& device;
+    std::vector<std::unique_ptr<btGatt::PrimaryService>>& services;
+  };
+  services_.clear();
+  Scope scope{*this, services_};
```

Comment:
This is in local scope that may not live when callback is called.

Reply:
`bt_gatt_client_foreach_services()` executes synchronously.

Reply:
I see, API signature kind of suggests that it's asynchronous, would you add a comment that it's synchronous?